include/swift/AST/DiagnosticEngine.h

```diff
@@ -629,7 +629,27 @@ namespace swift {
     DiagnosticState(DiagnosticState &&) = default;
     DiagnosticState &operator=(DiagnosticState &&) = default;
   };
-    
+
+  class LocalizationProducer {
+  public:
+    /// If the  message isn't available/localized in the current `yaml` file,
+    /// return the fallback default message.
+    virtual std::string getMessageOr(DiagID id,
```

Comment:
Let's make return of this function and `defualtMessage` be of type `StringRef` instead to avoid having to copy diagnostic around.

Reply:
Which means that `DiangosticEngine.h` would have to include `LocalizationFormat.h`